### PR TITLE
Optimize massive `fb_google_category_to_attribute_mapping.json` file

### LINKED ISSUE #1838: Very high memory usage when parsing `fb_google_category_to_attribute_mapping.json` (edit product page)
Originally discovered: https://github.com/woocommerce/facebook-for-woocommerce/pull/1810#discussion_r600985894

This json file is a massive 18.8MB in size and parsing it to an array uses ~100MB of memory on my machine. AFAICS this happens on the edit product screen but there might be other locations where it's used.

This memory usage is excessive and contributes to a high risk of memory exhaustion.

## Possible solutions

- **Optimize the file contents:** There seems to be a lot of duplicate content in the file maybe we can remove some of it.
- **Split up file contents based on category:** In the code, AFAICS, we are only reading one or two entries from the json file at a time so there's no reason we need to load the entire data set. Splitting up the contents by category is an option.
- **Convert the file contents to a more memory efficient format:** We could store it directly as a PHP array which _could_ reduce memory usage

### PR DESCRIPTION
**Pull request by danielbitzer**
Fixes #1838

This PR optimizes the file `fb_google_category_to_attribute_mapping.json` by removing large sequences of duplicate data. Most of the data in the file was actually duplicate attribute arrays. Instead of storing this duplicate data this PR extracts all attribute data into a separate look up file using a hashed references.

## The new files

- `google_category_to_attribute_mapping.json` this is the same as the original file but attribute arrays have been replaced with hashes
- `google_category_to_attribute_mapping_fields.json` this is a look-up file containing with the field hashes and the attribute data

## Memory reduction
On my machine I'm seeing ~15 MB to read these two .json files. Previously it was ~120 MB.

## Where is this code called
On product edit pages and category edit pages

## How to generate the json files
Run `npm run generate:category_attribute_json`

## Todo
- [x] some more testing
- [x] do some more data comparison 
- [x] add some more dev docs about this

### CODE REVIEW

**Comment by danielbitzer** [comment]
I was able to verify that the new data is identical to the original .json file by generating it from the methods and data added in this PR:

```php

$handler = facebook_for_woocommerce()->get_facebook_category_handler();

$cats = $handler->get_raw_attributes_data();
foreach( $cats as $cat_id => &$cat ) {
	// pull in attributes from separate file
	$cat = $handler->get_category_with_attrs( $cat_id );
}
$new_json_data = json_encode( $cats );


$original_contents = file_get_contents( facebook_for_woocommerce()->get_plugin_path() . '/bin/fb_google_category_to_attribute_mapping.json' );
// re-encode data for consistent json formatting
$original_json_data = json_encode( json_decode( $original_contents ));

var_dump( md5( $original_json_data ) );
var_dump( md5( $new_json_data ) );

```

The hashes were idential 🙌

File: bin/GenerateCategoryAttributeMapping.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
+<?php
+
+/**
+ * Class GenerateCategoryAttributeMapping
+ *
+ * Optimizes the 18 MB `fb_google_category_to_attribute_mapping.json` file by extracting duplicate attribute field data
+ * into two separate JSON files.
+ *
+ * `google_category_to_attribute_mapping.json` Is the same as the original file but attribute arrays have been replaced with hashes
+ * `google_category_to_attribute_mapping_fields.json` Is a look-up file containing with the attribute field array with hashes as keys.
```
Comment (budzanowski) [inline bin/GenerateCategoryAttributeMapping.php:10]:
```suggestion
 * `google_category_to_attribute_mapping_fields.json` Is a look-up file containing the attribute field array with hashes as keys.
```

**Review by budzanowski — APPROVED** [review]
Left a small nit but it looks really nice!

**Comment by haszari** [comment]
> On my machine I'm seeing ~15 MB to read these two .json files. Previously it was ~120 MB.

Nice work @danielbitzer !

**Comment by rcstr** [comment]
Hi @danielbitzer - Tested and works well

**Comment by danielbitzer** [comment]
Thank you all for the reviews! Because the `\SkyVerge\WooCommerce\Facebook\Products\FBCategories` class has undergone quite a few changes since I opened this PR I've had to rebase and tweak things a bit. 

The main change was the logic for checking parent category for attributes data if non was found on the current category (https://github.com/woocommerce/facebook-for-woocommerce/pull/1820). I have tested that this is all still working.

**Review by rcstr — APPROVED** [review]
I left a small suggestion, but overall looks great. I wonder if we should `deprecate` the function being renamed (`get_category_with_attrs`) and create a new one, I mention this because the function is public and _could_ be used by external developers.

**Comment by rcstr** [comment]
Tested running and worked as expected.

```php
add_action('init', function () {
	$handler = facebook_for_woocommerce()->get_facebook_category_handler();

	$cats = $handler->get_categories();
	foreach ($cats as $cat_id => &$cat) {
		// pull in attributes from separate file
		$cat = $handler->get_attributes_with_fallback_to_parent_category($cat_id);
	}
	$new_json_data = json_encode($cats);


	var_dump($new_json_data);
	exit();
}, 0);
```

File: bin/GenerateCategoryAttributeMapping.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+				$attr = $hash;
+			}
+
+			$this->category_export[ $category_id ] = $category;
+		}
+
+		file_put_contents( $plugin_root . $this::EXPORT_CATEGORIES_FILE_NAME, json_encode( $this->category_export ) );
+		file_put_contents( $plugin_root . $this::EXPORT_FIELDS_FILE_NAME, json_encode( $this->field_export ) );
+	}
```
[the remark this answers is not in the document]
Comment (danielbitzer) [reply bin/GenerateCategoryAttributeMapping.php]: Love it!

**Comment by danielbitzer** [comment]
Thanks @rcstr!

>  I wonder if we should deprecate the function being renamed (get_category_with_attrs) and create a new one, I mention this because the function is public and could be used by external developers.

Fair call! Added back and deprecated in 9c87df10e092ef8929295892186c9d310ec468ac.

**Comment by haszari** [comment]
@danielbitzer following up on this - I think this (or other changes in 2.4.0) fix these issues: #1680 #1700

Can you take a peek at those issues, and close if you think the issue is fixed (pretty sure it is!).

**Comment by danielbitzer** [comment]
Nice find! I actually stumbled across and closed a couple of other issues related to this last week but missed these.

**Pull request merged**